aggregator/src/aggregation/circuit.rs

```diff
@@ -0,0 +1,408 @@
+                    //
+                    let (assigned_aggregation_instances, acc) = aggregate::<Kzg<Bn256, Bdfg21>>(
+                        &self.svk,
+                        &loader,
+                        &self.snarks_with_padding,
+                        self.as_proof(),
+                    );
+                    log::trace!("aggregation circuit during assigning");
+                    for (i, e) in assigned_aggregation_instances[0].iter().enumerate() {
+                        log::trace!("{}-th instance: {:?}", i, e.value)
+                    }
+
+                    // extract the following cells for later constraints
+                    // - the accumulators
+                    // - the public inputs from each snark
+                    accumulator_instances.extend(flatten_accumulator(acc).iter().copied());
+                    // the snark is not a fresh one, assigned_instances already contains an
+                    // accumulator so we want to skip the first 12 elements from the public input
+                    snark_inputs.extend(
+                        assigned_aggregation_instances
+                            .iter()
+                            .flat_map(|instance_column| instance_column.iter().skip(ACC_LEN)),
+                    );
+
+                    config.range().finalize(&mut loader.ctx_mut());
+
+                    loader.ctx_mut().print_stats(&["Range"]);
+
+                    Ok((accumulator_instances, snark_inputs))
+                },
+            )?;
+
+            assert_eq!(snark_inputs.len(), MAX_AGG_SNARKS * DIGEST_LEN);
+            (accumulator_instances, snark_inputs)
+        };
+        end_timer!(timer);
+        // ==============================================
+        // step 2: public input aggregation circuit
+        // ==============================================
+        // extract all the hashes and load them to the hash table
+        let challenges = challenge.values(&layouter);
+
+        let timer = start_timer!(|| "load aux table");
+
+        #[cfg(not(feature = "disable_pi_aggregation"))]
+        let (hash_digest_cells, num_valid_snarks) = {
+            config
+                .keccak_circuit_config
+                .load_aux_tables(&mut layouter)?;
+            end_timer!(timer);
+
+            let timer = start_timer!(|| "extract hash");
+            // orders:
+            // - batch_public_input_hash
+            // - chunk\[i\].piHash for i in \[0, MAX_AGG_SNARKS)
+            // - batch_data_hash_preimage
+            let preimages = self.batch_hash.extract_hash_preimages();
+            assert_eq!(
+                preimages.len(),
+                MAX_AGG_SNARKS + 2,
+                "error extracting preimages"
+            );
+            end_timer!(timer);
+
+            let timer = start_timer!(|| ("assign hash cells").to_string());
+            let (hash_digest_cells, num_valid_snarks) = assign_batch_hashes(
+                &config,
```

Comment:
If we enabled the `skip_first_pass` feature, I think the aggregation circuit will fail again. To fix it, we can either 
1. allocate another set of columns for `assign_batch_hashes` or 
2. first return the `ctx` object in step 1 and pass it to `assign_batch_hashes`.